myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md

```diff
@@ -0,0 +1,662 @@
+        if full_sample:
+            idata_ar.extend(pm.sample(samples, random_seed=100, target_accept=0.95))
+            idata_ar.extend(pm.sample_posterior_predictive(idata_ar))
+        else:
+            return idata_ar
+
+    n = prediction_steps - ar_data.shape[0]
+
+    with AR:
+        # update values of predictors pass in nans for number of predicted steps
+        pm.set_data(
+            new_data={
+                "t": list(range(prediction_steps)),
+                "y": np.concatenate([ar_data, np.array([np.nan for i in range(n)])]),
+            },
+            coords={"obs_id": list(range(prediction_steps))},
+        )
+        # use the updated values and predict outcomes and probabilities:
+        idata_preds = pm.sample_posterior_predictive(
+            idata_ar, var_names=["likelihood"], predictions=True, random_seed=100
+        )
+
+    return idata_ar, idata_preds, AR
+```
+
+We will fit this model by specifying priors on the negative trend and the range of the standard deviation to respect the direction of the data drift.
+
+```{code-cell} ipython3
+priors_0 = {
+    "coef_0": {"mu": 0.2, "sigma": 0.5, "size": 2},
+    "coef_1": {"mu": 0.2, "sigma": 0.03, "size": 2},
+    "alpha": {"mu": -4, "sigma": 0.1},
+    "beta": {"mu": -0.5, "sigma": 0.1},
+    "sigma": 8,
+    "init": {"mu": -4, "sigma": 0.1, "size": 1},
+}
+
+
+idata_trend, preds_trend, model = make_latent_AR_trend_model(y_t, priors_0, full_sample=True)
+```
+
+```{code-cell} ipython3
+pm.model_to_graphviz(model)
+```
+
+We can see the structure more clearly with the plate notation, and this additional structure has helped to appropriately fit the directional trend of the timeseries data.
+
+```{code-cell} ipython3
+plot_fits(idata_trend, preds_trend);
+```
+
+```{code-cell} ipython3
+az.summary(idata_trend, var_names=["coef_0", "coef_1", "sigma", "alpha", "beta"])
+```
+
+## Complicating the picture further
+
+Next we'll add a seasonal component to our data and see how we can recover this aspect of the data with a bayesian structural timeseries model.
+
+```{code-cell} ipython3
+t_data = list(range(200))
+n_order = 10
+periods = np.array(t_data) / 7
+
+fourier_features = pd.DataFrame(
+    {
+        f"{func}_order_{order}": getattr(np, func)(2 * np.pi * periods * order)
+        for order in range(1, n_order + 1)
+        for func in ("sin", "cos")
+    }
+)
+
+y_t_s = y_t + 20 * fourier_features["sin_order_1"]
+
+fig, ax = plt.subplots(figsize=(10, 4))
+ax.plot(y_t_s)
+ax.set_title("AR + Trend + Seasonality");
+```
+
+The key to fitting this model is to understand that we're now passing in synthetic fourier features and that the model now expects a linear combination of those synthetic features even in the prediction step. As such we need to be able to supply those features even out into the future. This fact remains key for any other type of predictive feature we might want to add e.g. day of the week, holiday flag or any other. 
```

Comment:
I find this paragraph very hard to read. I understand what you want to say, but maybe you could pause a bit more to give a longer explanation of what it is that you’re doing. Imagine that many readers won’t know what a Fourier decomposition is, so maybe you can give some intuition behind (without going into the math though)

Reply:
I've re-written this paragraph and expanded it out a bit more (hopefully clearer).